File: EM and NLP/src/main/java/MDP.java
import com.google.common.base.Stopwatch;
import org.apache.commons.lang3.StringUtils;

import java.io.*;
import java.util.ArrayList;
import java.util.concurrent.Callable;
import java.util.concurrent.TimeUnit;

import static java.lang.Math.pow;
import static java.lang.Math.random;

/**
 * Created by tarun on 4/6/17.
 */
public class MDP implements Callable<Void> {

    int T;
    ArrayList<Integer> locs;
    int nlocs;
    int agent;
    ArrayList<State> states;
    ArrayList<Action> actions;
    private int[] collectTimes;
    private int [][] transitTimes;
    double alpha;
    double beta;
    ArrayList<Double> start;
    ArrayList<String> binary;
    private Config config;
    State terminal;
    private int numberStates;
    int numberActions;
    int numberVariables;

    MDP(int agent, Config c) {
        this.T = c.T[agent];
        this.locs = c.locs.get(agent);
        this.nlocs = locs.size();
        this.agent = agent;
        this.binary = new ArrayList<>();
        generateBinaryStrings(this.nlocs);
        this.states = new ArrayList<>();
        this.actions = new ArrayList<>();
        this.collectTimes = c.collectTimes[agent];
        this.transitTimes = c.transitTimes[agent];
        this.alpha = c.alpha;
        this.beta = c.beta;
        this.start = new ArrayList<>();
        this.config = c;
    }

    @Override
    public Void call() {
        if (this.config.flag==0) {
            generateDomain();
        } else {
            deserializeDomain();
            defineStart();
            this.numberStates = states.size();
            this.numberActions = actions.size();
            this.numberVariables = numberStates * numberActions;
        }
        return null;
    }

    private void defineStart() {
        double sum = 0;
        for(State s : states) {
            boolean sameds = true;
            for (int i = 0; i < nlocs; i++) {
                if (s.dvals.charAt(i) != '0') {
                    sameds = false;
                    break;
                }
            }

            if(s.time==config.T[agent] && sameds && s.dold==0) {
                sum += 1;
            }
        }

        for(State s : states) {
            boolean sameds = true;
            for (int i = 0; i < nlocs; i++) {
                if (s.dvals.charAt(i) != '0') {
                    sameds = false;
                    break;
                }
            }
            if(s.time==config.T[agent] && sameds && s.dold==0) {
                start.add((double)1/sum);
            } else {
                start.add((double)0);
            }
        }


    }

    private void generateDomain() {
        String dum= StringUtils.repeat('2', nlocs);
        this.terminal = new State(0,-1,-1,-1, dum, -1, actions);
        this.states.add(this.terminal);
        this.initiateActions();
        this.initiateStates();
        //checkTransitionProbabilitySumTo1();
        this.waste();
        this.reindexStates();
        this.writeTransitions();
        this.writeRewards();
        this.serializeDomain();
        this.defineStart();
        this.numberStates = states.size();
        this.numberActions = actions.size();
        this.numberVariables = numberStates * numberActions;
    }

    ArrayList<double[]> generateLPAc() {
        System.out.println("Generating LP for "+agent);
        double Rmax = config.Rmax;
        double Rmin = config.Rmin;
        ArrayList<double[]> ret = new ArrayList<>();
        double[] R_mat = new double[numberVariables];
        double[] newR = new double[numberVariables];

        int store = 0;
        for(State s : states) {
            for(Reward rew : s.reward) {
                R_mat[store] = rew.reward;
                newR[store] = (rew.reward / (Rmax-Rmin));
                store += 1;
            }
        }

        ret.add(R_mat);
        ret.add(newR);
        return ret;
    }

    private void writeTransitions() {
        System.out.println("     Writing Transitions for Agent " +agent);
        for(Action a : actions) {
            for(State s : states) {
                for(State sd : states) {
                    double tt = transition(s,a,sd);
                    if (tt != 0) {
                        s.transition.add(new Transition(a.index, sd.index, tt));
                    }
                }
            }
        }
    }

    private void writeRewards() {
        System.out.println("     Writing Rewards for Agent " +agent);
        for(State s : states) {
            for(Action a : actions) {
                double reward = rewardFunction(s,a);
                s.reward.add(new Reward(a.index, reward));
            }
        }
    }

    private void reindexStates() {
        int index = 0;
        for (State s:states) {
            s.index = index;
            index += 1;
        }
    }

    private void serializeDomain() {
        try {
            FileOutputStream fos = new FileOutputStream(this.config.workDir + "Data/Domain" + agent + "_exp_" + config.experiment + ".ser");
            ObjectOutputStream oos = new ObjectOutputStream(fos);
            oos.writeObject(states);
            oos.writeObject(actions);
            for(State s: states) {
                oos.writeObject(s.transition);
                oos.writeObject(s.reward);
            }
            oos.flush();
            oos.close();
            fos.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void deserializeDomain()    {
        try {
            FileInputStream fos = new FileInputStream(this.config.workDir + "Data/Domain" + agent + "_exp_" + config.experiment + ".ser");
            ObjectInputStream oos = new ObjectInputStream(fos);
            states = (ArrayList<State>) oos.readObject();
            terminal = states.get(0);
            actions = (ArrayList<Action>) oos.readObject();
            for(State s : states) {
                s.transition = (ArrayList<Transition>) oos.readObject();
                s.reward = (ArrayList<Reward>) oos.readObject();
            }
            oos.close();
            fos.close();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
    }

    void checkTransitionProbabilitySumTo1() {
        for (Action k : actions) {
            for (State i : states) {
                double sum = 0;
                for (State j : states) {
                    double tran = transition(i,k,j);
                    sum += tran;
                }
                if (sum!=1) {
                    System.out.println("WARNING: k: " + k + " i: " + i + " Sum: " + sum);
                }
            }
        }
    }

    private void generateBinaryStrings(int n) {
        for(int i = 0; i < pow(2,n); i++) {
            String form="%0"+n+"d";
            String bin = String.format(form, Integer.valueOf(Integer.toBinaryString(i)));
            binary.add(bin);
        }
    }

    private void initiateActions() {
        int index = 0;
        Action a = new Action(index, "Collect");
        this.actions.add(a);
        index += 1;
        for (int i = 0; i < nlocs ; i++) {
            a = new Action(index, "Go to Site "+i, i);
            index += 1;
            this.actions.add(a);
        }
    }

    private void initiateStates() {
        int index = 1;
        for (int i = 0; i <= this.T; i+= config.collectTimes[agent][0]) {
            for (int j = 0; j < nlocs ; j++) {
                for (String k:binary) {
                    ArrayList<Integer> lyst = new ArrayList<>();
                    if (k.charAt(j)=='0') {
                        lyst.add(0);
                    } else if (k.charAt(j)=='1') {
                        lyst.add(0);
                        lyst.add(1);
                    }
                    for (Integer t:lyst) {
                        State st = new State(index,j,locs.get(j),i,k,t,actions);
                        this.states.add(st);
                        index += 1;
                    }
                }
            }
        }
        System.out.println(states.size() * actions.size());
    }

    private double rewardFunction(State s, Action a) {
        if (s.equals(terminal)) {
            return 0.0;
        }
        int sdvalsloc = Character.getNumericValue(s.dvals.charAt(s.location));
        if(s.dold == 0 && sdvalsloc == 1) {
            return config.rewardCollection[agent][s.location] + random();
        } else {
            return 0.1 + random();
        }
    }

    double transition(State s, Action a, State sd) {
        if (a.name.equals("Collect")) {
            return transitionCollect(s,sd);
        } else {
            return transitionGoto(s,sd,a);
        }
    }

    private double transitionCollect(State s, State sd) {
        if (s.time == -1 && sd.equals(s)){
            return 1.0;
        }

        if (s.time==0 && sd.equals(terminal)) {
            return 1.0;
        }

        boolean sameds = true;
        for (int i = 0; i < nlocs; i++) {
            if (i != s.location) {
                if (s.dvals.charAt(i)!=sd.dvals.charAt(i)) {
                    sameds = false;
                    break;
                }
            }
        }

        if (!sameds) {
            return 0.0;
        }

        if (s.location != sd.location) {
            return 0.0;
        }

        if (sd.time != s.time-this.collectTimes[s.location]) {
            return 0.0;
        }

        if (sd.time < 0) {
            return 0.0;
        }

        int sdvalsloc = Character.getNumericValue(s.dvals.charAt(s.location));
        int sddvalsloc = Character.getNumericValue(sd.dvals.charAt(s.location));
        if (sd.dold != sdvalsloc) {
            return 0.0;
        }

        if (s.dold==1 && sdvalsloc==0) {
            return 0.0;
        }

        if (sdvalsloc==0) {
            return (sddvalsloc==1)?alpha:(1-alpha);
        } else if (sdvalsloc==1) {
            return (sddvalsloc==1)?1.0:0.0;
        } else {
            return 0.0;
        }
    }

    private double transitionGoto(State s, State sd, Action a) {
        int l = s.location;
        int ld = sd.location;
        int t = s.time;
        int td = sd.time;

        int dold = s.dold;
        int doldd = sd.dold;
        int dest = a.gotox;

        boolean sameds = true;
        for (int i = 0; i < nlocs; i++) {
            if (s.dvals.charAt(i) != sd.dvals.charAt(i)) {
                sameds = false;
                break;
            }
        }

        if (s.time==-1 && sd.equals(s)) {
            return 1.0;
        }

        if (s.time==0 && sd.equals(terminal)) {
            return 1.0;
        }

        if (!sameds) {
            return 0.0;
        }

        if ((ld != dest) && (ld != l)) {
            return 0.0;
        }

        if (td != t - transitTimes[l][ld]) {
            return 0.0;
        }

        if (td < 0) {
            return 0.0;
        }

        int sdvalsloc = Character.getNumericValue(s.dvals.charAt(s.location));
        int sddvalsloc = Character.getNumericValue(sd.dvals.charAt(sd.location));

        if (doldd != sddvalsloc) {
            return 0.0;
        }

        if (dold==1 && sdvalsloc==0) {
            return 0.0;
        }

        if (ld==l && ld==dest) {
            return 1.0;
        } else if (ld==l) {
            return 1 - beta;
        } else if (ld==dest) {
            return beta;
        } else {
            return 0.0;
        }
    }

    private ArrayList<State> removeWasteStates(int iter) {
        ArrayList<State> wasteStates  = new ArrayList<>();
        int sum = 0;
        double val = 0.0;
        Stopwatch timer = Stopwatch.createUnstarted();
        timer.start();
        int tots = states.size();
        int offset = config.offset;
        for (State s:states) {
            sum += 1;
            if (sum%offset == 0) {
                timer.stop();
                //System.out.println(timer);
                val = (double)timer.elapsed(TimeUnit.SECONDS);
                double ntimes = (double)sum / (double)offset;
                double avg = val / ntimes;
                double timerem = ((double)(tots-sum) / offset) * avg;
                System.out.println("["+agent+","+iter+"] Done. "+sum+" Out of: "+tots+ " Avg: "+avg + " Rem: "+timerem);
                timer.start();
            }

            if (s.equals(terminal)) {
                //System.out.println("sasassas");
                continue;
            }

            int num_value = Character.getNumericValue(s.dvals.charAt(s.location));
            if (s.dold==1 && num_value==0) {
                System.out.print("10 Anomaly");
                wasteStates.add(s);
                continue;
            }

            boolean sameds = true;
            for (int i = 0; i < nlocs; i++) {
                if (s.dvals.charAt(i) != '0') {
                    sameds = false;
                    break;
                }
            }

            if (s.time==T && sameds && s.dold==0) {
                continue;
            }

            int flag = 0;
            for(State sd : states) {
                for (Action a :  actions) {
                    //System.out.println(sd + "\n" + a + "\n" + s + "\n");
                    if (transition(sd,a,s)!=0) {
                        flag = 1;
                        break;
                    }
                }
                if (flag==1) {
                    break;
                }
            }

            if (flag==0) {
                wasteStates.add(s);
            }
        }
        timer.reset();
        states.removeAll(wasteStates);
        System.out.println("For agent " + agent+ " Iter "+iter+" done and removed "+wasteStates.size()+".");
        return wasteStates;
    }

    private ArrayList<State> succRemoval(ArrayList<State> removedSt, int iter) {
        ArrayList<State> wasteStates = new ArrayList<>();
        ArrayList<State> maybe = new ArrayList<>();
        wasteStates.clear();
        maybe.clear();

        int sum = 0;
        double val = 0.0;
        Stopwatch timer = Stopwatch.createUnstarted();
        timer.start();
        int tots = removedSt.size();
        int offset = config.offset;
        for (State rms:removedSt) {
            sum += 1;
            if (sum%offset == 0) {
                timer.stop();
                //System.out.println(timer);
                val = (double)timer.elapsed(TimeUnit.SECONDS);
                double ntimes = (double)sum / (double)offset;
                double avg = val / ntimes;
                double timerem;
                timerem = ((double)(tots-sum) / offset) * avg;
                System.out.println("["+agent+","+iter+"] Done. "+sum+" Out of: "+tots+ " Avg: "+avg + " Rem: "+timerem);
                timer.start();
            }

            for(State sts : states) {

                if (sts.equals(terminal)) {
                    continue;
                }

                int num_value = Character.getNumericValue(sts.dvals.charAt(sts.location));
                if (sts.dold==1 && num_value==0) {
                    System.out.print("10 Anomaly");
                    wasteStates.add(sts);
                    continue;
                }

                boolean sameds = true;
                for (int i = 0; i < nlocs; i++) {
                    if (sts.dvals.charAt(i) != '0') {
                        sameds = false;
                        break;
                    }
                }

                if (sts.time==T && sameds && sts.dold==0) {
                    continue;
                }

                for (Action a :  actions) {
                    if (transition(rms,a,sts)!=0) {
                        maybe.add(sts);
                        break;
                    }
                }
            }
        }

        for (State sts : maybe) {
            int flag = 0;
            for (State s : states) {
                for (Action a : actions) {
                    if (transition(s,a,sts)!=0.0) {
                        flag=1;
                        break;
                    }
                }
                if (flag==1) {
                    break;
                }
            }
            if (flag==0) {
                wasteStates.add(sts);
            }
        }

        ArrayList<State> prevs = new ArrayList<>();
        for (State torem : wasteStates) {
            if(states.contains(torem)) {
                prevs.add(torem);
                states.remove(torem);
            }
        }
        System.out.println("For agent " + agent+ " Iter "+iter+" done and removed "+prevs.size()+".");
        return prevs;
    }

    private void waste() {
        int iter = 1;
        ArrayList<State> removed = removeWasteStates(iter);
        while (removed.size() != 0) {
            iter += 1;
            removed = succRemoval(removed, iter);
        }
    }
}

File: RL/Part-I/src/MDP.py
from State import State
from Action import Action
import itertools
import random
import numpy as np
import math
import time
import pickle

class MDP:

    def __init__(self, agent, config):
        self.T = config.T
        self.line = config.agentWiseLines[agent]
        self.locs = self.line.getLocations()
        self.nlocs = self.line.getNumberOfLocations()
        self.agent = agent
        self.lst = list(itertools.product([0, 1], repeat=self.nlocs))
        self.states = []
        self.actions = []
        self.actionTime = config.actionTimes[agent]
        self.alpha = config.alpha
        self.beta = config.beta
        self.repeatEveryKTimeSteps = config.repeatKTimeSteps
        self.start = []
        self.policy = None
        self.terminal = None
        self.numberStates = None
        self.numberActions = None
        self.numberVariables = None
        self.config = config

    def generateDomain(self):
        self.terminal = State(0, None, None, None, None, None)
        self.states.append(self.terminal)
        self.initiateActions()
        self.initiateStates()
        self.waste()
        self.reindexStates()
        self.checkTransitionProbabilitySumTo1()
        self.writeTransitions()
        self.writeRewards()
        self.serializeDomain()
        self.defineStart()
        self.numberStates = len(self.states)
        self.numberActions = len(self.actions)
        self.numberVariables = self.numberStates * self.numberActions

    def initiateActions(self):
        self.actions.append(Action(0, "Inspect"))
        self.actions.append(Action(1, "Next"))
        self.actions.append(Action(2, "Previous"))

    def initiateStates(self):
        index = 1
        for i in xrange(0, self.T + 1, self.actionTime):
            for j in xrange(0, self.nlocs):
                for k in self.lst:
                    if k[j] == 0:
                        lyst = [0]
                    elif k[j] == 1:
                        lyst = [0, 1]
                    for t in lyst:
                        st = State(index, j, i, k, t, self.actions)
                        self.states.append(st)
                        index = index + 1

    def transition(self, s, a, sd):

        if s == self.terminal:
            if sd == s:
                return 1
            else:
                return 0

        if s.time == 0:
            if sd == self.terminal:
                return 1
            else:
                return 0

        if sd == self.terminal:
            return 0.0

        if a.name == "Inspect":
            return self.transitionInspect(s, sd)
        else:
            return self.transitionMove(s, sd, a)

    def waste(self):
        iter = 1
        removed = self.removeWasteStates(iter)
        while len(removed) != 0:
            iter += 1
            removed = self.succRemoval(removed, iter)

    def removeWasteStates(self, iter):
        wastestates = []
        sum = 0.0
        val = 0.0
        start = time.time()
        tots = len(self.states)
        offset = 500
        for i in self.states:
            sum += 1
            if sum % offset == 0:
                end = time.time()
                val += float(end - start)
                ntimes = float(sum) / float(offset)
                avg = float(val) / float(ntimes)
                timerem = (float(tots - sum) / float(offset)) * avg
                print "Done. " + str(sum) + " Out of: " + str(tots) + " Avg: " + str(avg) + " Rem: " + str(timerem)
                start = time.time()

            if i == self.terminal:
                continue

            if i.dold == 1 and i.dvals[i.location] == 0:
                wastestates.append(i)
                continue

            sameds = all([i.dvals[j] == 0 for j in xrange(0, self.nlocs)])
            if i.time == self.T and sameds == True and i.dold == 0:
                continue

            flag = 0
            for j in self.states:
                for k in self.actions:
                    if self.transition(j, k, i) != 0:
                        flag = 1
                        break
                if flag == 1:
                    break
            if flag == 0:
                wastestates.append(i)

        for x in wastestates:
            self.states.remove(x)

        print "Iter " + str(iter) + " done and removed " + str(len(wastestates)) + "."
        return wastestates

    def succRemoval(self, removedSt, iter):

        wasteStates = []
        maybe = []

        sumd = 0
        val = 0.0
        start = time.time()
        tots = len(removedSt)
        offset = 500
        for rms in removedSt:

            sumd += 1
            if sumd % offset == 0:
                end = time.time()
                val += float(end - start)
                ntimes = float(sumd) / float(offset)
                avg = float(val) / float(ntimes)
                timerem = (float(tots - sumd) / float(offset)) * avg
                print "Done. " + str(sumd) + " Out of: " + str(tots) + " Avg: " + str(avg) + " Rem: " + str(timerem)
                start = time.time()

            for sts in self.states:

                if sts == self.terminal:
                    continue

                if sts.dold == 1 and sts.dvals[sts.location] == 0:
                    wasteStates.append(sts)
                    continue

                sameds = all([sts.dvals[j] == 0 for j in xrange(0, self.nlocs)])
                if sts.time == self.T and sameds == True and sts.dold == 0:
                    continue

                for a in self.actions:
                    if self.transition(rms, a, sts) != 0:
                        maybe.append(sts)
                        break

        for i in maybe:
            flag = 0
            for j in self.states:
                for k in self.actions:
                    if self.transition(j, k, i) != 0:
                        flag = 1
                        break
                if flag == 1:
                    break
            if flag == 0:
                wasteStates.append(i)

        prevs = []
        for torem in wasteStates:
            if torem in self.states:
                prevs.append(torem)
                self.states.remove(torem)

        print "Iter " + str(iter) + " done and removed " + str(len(prevs)) + "."
        return prevs

    def checkTransitionProbabilitySumTo1(self):
        for k in self.actions:
            for i in self.states:
                sum = 0
                for j in self.states:
                    tran = self.transition(i, k, j)
                    sum += tran
                if (sum != 1):
                    print "WARNING: k: " + str(k) + " i: " + str(i) + " Sum: " + str(sum)

    def transitionInspect(self, s, sd):

        if s.location != sd.location:
            return 0

        if sd.time != s.time - self.actionTime:
            return 0

        if sd.time < 0:
            return 0

        sameds = all([s.dvals[j] == sd.dvals[j] for j in xrange(0, self.nlocs) if j != s.location])

        if (self.isReset(s)):
            sameds = all(sd.dvals[j] == 0 for j in xrange(0, self.nlocs) if j != sd.location)

        if sameds == False:
            return 0

        if sd.dold != s.dvals[s.location]:
            return 0

        if s.dold == 1 and s.dvals[s.location] == 0:
            return 0

        if s.dvals[s.location] == 0:
            return self.alpha if sd.dvals[sd.location] == 1 else (1 - self.alpha)
        elif s.dvals[s.location] == 1:
            return 1 if sd.dvals[sd.location] == 1 else 0
        else:
            return 0

    # defines the next location agent can travel to.
    def getNextLocation(self, location):
        if (location + 1) < self.nlocs:
            return location + 1
        else:
            return location

    # defines previous location agent can travel to.
    def getPreviousLocation(self, location):
        if (location - 1) >= 0:
            return location - 1
        else:
            return location

    def transitionMove(self, s, sd, action):

        l = s.location
        ld = sd.location
        t = s.time
        td = sd.time

        dold = s.dold
        doldd = sd.dold

        dest = None
        if (action.name == "Next"):
            dest = self.getNextLocation(s.location)
        elif (action.name == "Previous"):
            dest = self.getPreviousLocation(s.location)

        if (dest is None):
            return 0.0

        sameds = all([s.dvals[j] == sd.dvals[j] for j in xrange(0, self.nlocs)])

        if (self.isReset(s)):
            sameds = all(sd.dvals[j] == 0 for j in xrange(0, self.nlocs))

        if sameds == False:
            return 0

        if (ld != dest and ld != l):
            return 0

        if td != t - self.actionTime:
            return 0

        if td < 0:
            return 0

        if doldd != sd.dvals[ld]:
            return 0

        if dold == 1 and s.dvals[l] == 0:
            return 0

        if ld == l and ld == dest:
            return 1.0
        elif ld == l:
            return 1 - self.beta
        elif ld == dest:
            return self.beta
        else:
            return 0.0

    def rewardFunction(self, s, a):
        if s == self.terminal:
            return 0.0
        if s.dold == 0 and s.dvals[s.location] == 1:
            return self.config.rewardCollection[self.agent][s.location]
        else:
            return 0.0

    def call(self):
        import datetime
        print "Start: ",datetime.datetime.now()
        if self.config.flag == 0:
            self.generateDomain()
        else:
            self.deserializeDomain()
            self.defineStart()
            self.numberStates = len(self.states)
            self.numberActions = len(self.actions)
            self.numberVariables = self.numberStates * self.numberActions
        print "End: ",datetime.datetime.now()
        return self

    def reindexStates(self):
        index = 0
        lst = []
        for x in self.states:
            a = State(index, x.location, x.time, x.dvals, x.dold, self.actions)
            a.setReward(x.reward)
            a.setTransition(x.transition)
            lst.append(a)
            index = index + 1
        self.states = lst

    def serializeDomain(self):
        f = open(self.config.workDir + "Data/Domain" + str(self.agent) + "_exp_" + str(self.config.experiment) + ".pickle", 'w')
        pickle.dump(self.states, f)
        pickle.dump(self.actions, f)
        for s in self.states:
            pickle.dump(s.transition, f)
            pickle.dump(s.reward, f)
        f.close()

    def deserializeDomain(self):
        f = open(self.config.workDir + "Data/Domain" + str(self.agent) + "_exp_" + str(self.config.experiment) + ".pickle")
        self.states = pickle.load(f)
        self.terminal = self.states[0]
        self.actions = pickle.load(f)
        for s in self.states:
            s.transition = pickle.load(f)
            s.reward = pickle.load(f)
        f.close()

    def defineStart(self):
        sum = 0
        for i in self.states:
            if i==self.terminal:
                continue
            sameds = all([i.dvals[j] == 0 for j in xrange(0, self.nlocs)])
            if i.time == self.T and sameds == True and i.dold == 0:
                sum += 1
        for i in self.states:
            if i==self.terminal:
                self.start.append(float(0))
                continue
            sameds = all([i.dvals[j] == 0 for j in xrange(0, self.nlocs)])
            if i.time == self.T and sameds == True and i.dold == 0:
                self.start.append(float(1 / float(sum)))
            else:
                self.start.append(float(0))

    def isReset(self, state):
        curTime = state.time
        remTime = curTime - self.actionTime
        if (remTime >= 0):
            n = (float(self.T) - float(remTime)) / float(self.repeatEveryKTimeSteps)
            int_n = int(n)
            if (int_n == n):
                return True
        return False

    def writeTransitions(self):
        print "     Writing Transitions for Agent " + str(self.agent)
        for a in self.actions:
            for s in self.states:
                for sd in self.states:
                    tt = self.transition(s, a, sd)
                    if tt != 0:
                        s.transition.append((a.index, sd.index, tt))

    def writeRewards(self):
        print "     Writing Rewards for Agent " + str(self.agent)
        for s in self.states:
            for a in self.actions:
                reward = self.rewardFunction(s, a)
                s.reward.append((a.index, reward))

    def generateLPAc(self, gamma, genA=False):
        print "Generating LP "
        A_mat = []
        if genA is True:

            decisionvar = []
            for x in self.states:
                triple = []
                for y in self.states:
                    triplet = []
                    for a in y.possibleActions:
                        if x.index == y.index:
                            triplet.append(float(1))
                        else:
                            triplet.append(float(0))
                    triple.append(triplet)
                decisionvar.append(triple)

            for x in self.states:
                incoming = []
                for s in self.states:
                    for t in s.transition:
                        if t[1] == x.index and t[2] != 0:
                            incoming.append((s, t[0], t[2]))

                for h in incoming:
                    decisionvar[x.index][h[0].index][h[1]] -= gamma * float(h[2])

            A_mat = []
            for x in decisionvar:
                lit = []
                for t in x:
                    lit.extend(t)
                A_mat.append(lit)

        R_mat = []
        for x in self.states:
            for y in x.possibleActions:
                for r in x.reward:
                    if r[0] == y.index:
                        R_mat.append(r[1])
        return np.array(A_mat), np.array(R_mat)

    def solveLP(self, gamma):
        A, R = self.generateLPAc(gamma, genA=True)
        print np.shape(A)
        print np.shape(R)

        R_mat = np.array(R)[np.newaxis].T
        A_mat = np.array(A)

        alpha = self.start

        x = cvxpy.Variable(self.numberVariables, 1)
        obj = cvxpy.Maximize(np.transpose(R_mat)*x)
        constraints = [A_mat*x == alpha, x >= 0.00001]
        prob = cvxpy.Problem(obj, constraints)
        prob.solve()
        #print "status:", prob.status
        print "LPsolver: optimal value", prob.value
        #print "Optimal x: ", x.value
        print "Sum of x values: ", cvxpy.sum_entries(x).value
        return x.value

    def getOptimalPolicy(self, gamma):
        self.policy = dict()
        xvals = self.solveLP(gamma)
        for s in self.states:
            xs = [xvals[(s.index * self.numberActions) + a.index] for a in self.actions]
            xs_sum = sum(xs)
            probs = [float(val)/float(xs_sum) for val in xs]
            assert sum(probs) - 1 <= 1e-5
            self.policy[s.index] = probs
        return self.policy

    def hasStateWith(self, location, time, dvals, dold):
        for s in self.states:
            if s.location == location and s.time == time and all(s.dvals[vals] == dvals[vals] for vals in xrange(0, len(dvals))) and s.dold == dold:
                assert s != self.terminal
                return s.index
        return False

    def hasStateWithActualLocation(self, actlocation, time, dvals, dold):
        ind = self.line.getIndexOfLocation(actlocation)
        assert ind != -1
        for s in self.states:
            if s.location == ind and s.time == time and all(s.dvals[vals] == dvals[vals] for vals in xrange(0, len(dvals))) and s.dold == dold:
                assert s != self.terminal
                return s.index
        return False
